File: day25/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type State struct {
	write [2]int
	move  [2]int
	next  [2]string
}

var (
	tape         = make(map[int]int)
	cursor       int
	state        string
	stateTrans   = make(map[string]State)
	steps, count int
)

func main() {
	file, _ := os.Open("day25/input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Scan() // Read the initial state line
	state = strings.Split(scanner.Text(), " ")[3][:1]
	scanner.Scan() // Read the steps line
	stepsStr := strings.Split(scanner.Text(), " ")[5]
	steps, _ = strconv.Atoi(stepsStr)
	for scanner.Scan() {
		parseInput(scanner)
	}
	for i := 0; i < steps; i++ {
		execute()
	}
	for _, v := range tape {
		count += v
	}
	fmt.Println(count)
}

func parseInput(scanner *bufio.Scanner) {
	line := scanner.Text()
	if strings.HasPrefix(line, "In state ") {
		stateName := strings.Split(line, " ")[2][:1]
		var trans State
		for i := 0; i < 2; i++ {
			scanner.Scan() // Read the value condition line
			scanner.Scan() // Read the write action line
			trans.write[i], _ = strconv.Atoi(strings.Split(scanner.Text(), " ")[4][:1])
			scanner.Scan() // Read the move action line
			moveDirection := strings.Split(scanner.Text(), " ")[6]
			if moveDirection == "right." {
				trans.move[i] = 1
			} else {
				trans.move[i] = -1
			}
			scanner.Scan() // Read the next state line
			trans.next[i] = strings.Split(scanner.Text(), " ")[4][:1]
		}
		stateTrans[stateName] = trans
	}
}

func execute() {
	trans := stateTrans[state]
	tapeVal := tape[cursor]
	tape[cursor] = trans.write[tapeVal]
	cursor += trans.move[tapeVal]
	state = trans.next[tapeVal]
}

File: day25/part1.py

import re

def parse_input(file_path):
    with open(file_path, 'r') as file:
        lines = file.read().splitlines()
    initial_state = lines[0][-2]
    steps = int(re.findall(r'\d+', lines[1])[0])
    states = {}
    for i in range(3, len(lines), 10):
        state = lines[i][-2]
        value_0 = int(lines[i+2][-2])
        move_0 = -1 if lines[i+3].split()[-1] == 'left.' else 1
        next_state_0 = lines[i+4][-2]
        value_1 = int(lines[i+6][-2])
        move_1 = -1 if lines[i+7].split()[-1] == 'left.' else 1
        next_state_1 = lines[i+8][-2]
        states[state] = {(0,): (value_0, move_0, next_state_0), (1,): (value_1, move_1, next_state_1)}
    return initial_state, steps, states

def run_turing_machine(file_path):
    state, steps, states = parse_input(file_path)
    tape = {}
    cursor = 0
    for _ in range(steps):
        value = tape.get(cursor, 0)
        new_value, move, next_state = states[state][(value,)]
        tape[cursor] = new_value
        cursor += move
        state = next_state
    checksum = sum(tape.values())
    return checksum

print(run_turing_machine("day25/input.txt"))
